### ISSUE TITLE
Make shell configurable

### ISSUE THREAD

**Issue by WindSoilder** [Nukesor/pueue#363]
Hi, thank you so much for your great work, after using it for a while I found that I fell in love with it :-D I'm using it to schedule my background tasks.

#### Is your feature request related to a problem?
Currently I'm trying to make [nushell](https://www.nushell.sh/) supports background task, and I want to implemented it based on `pueued` and `pueue-lib` :)

After investigating code, I found that pueued runs `sh -c xxx` on linux/mac, and runs `powershell -c xxx` on windows.  It would be awesome to support a configuration point where other shells, like nushell, could be configured to be used instead of only bash and powershell. For instance, I'd like to call pueue with `nu -c xxx`, but I cannot right now.

#### Describe the solution you'd like

Make internal shell configurable while running pueue and pueued.  An example would be something like this:
```yaml
shared:
  shell: nu
```

If the `shell` configuration doesn't exists, make it as default like currently.

#### Describe alternatives you've considered

None

#### Additional context

I have changed one line by myself for experimental:
```rust
pub fn compile_shell_command(command_string: &str) -> Command {
    let mut command = Command::new("nu");  // changed line
    command.arg("-c").arg(command_string);

    command
}
```

And running pueue, I get the following result:
<img width="881" alt="Screen Shot 2022-09-16 at 22 40 40" src="https://user-images.githubusercontent.com/22256154/190665463-50571ba4-a606-47c7-a859-f7cacd3b911e.png">

Additionally, if we support this, we can even support adding running powershell commands on linux/mac, because powershell supports many platform for now (https://github.com/PowerShell/PowerShell#get-powershell)

One thing to consider is environment transfer to `pueued`, I'm considering build a client based on `pueue-lib`(because for querying job status, I need to generate a datastructure that nushell used, it can be implemented there)

**Comment by Nukesor** [comment]
First of all, thanks for the feature request.

*** Disclaimer start ***
A small upfront disclaimer:
Whenever I get feature requests for significant changes in one of my projects, my first comment is usually quite negative.
This is due to the fact, that I have to assess the potential negative consequences this change might have.
The main reasons for this are
- to prevent feature creep
- to keep it maintainable by a single person
- to keep it user friendly
- to keep it as stand-alone as possible

This doesn't mean that I'm against implementing this feature.
I got convinced of adding a feature more than once.
More than once, features were less problematic than I expected.

The following is a unfiltered response to your proposal written 15 minutes after reading it ;D

*** Disclaimer end ***

I already had the shell configurable in the first version of pueue, which was written in Python.

Back then, it actually made quite a lot of problems and lead to many bug reports due to various shells not working as expected in a non-shell environment.
One of the main problems was that a lot of shell configuration logic and some plugins caused weird behavior and/or error messages showing up in the log output.

In general, I like the idea of using the same shell I as my daily driver.
However, I fear that this will lead to a flood of issues and future problems, as this will allow users to freely set Pueue's main process executor.

## Process handling

The current subprocess handling logic is very much optimized to work with `sh` and accounts for a few behaviors one might not expect.
For instance, `['sh', '-c', 'some command']` actually doesn't always spawn a `sh` process.
In fact, it only spawns an actual `sh -c` process if there's some shell logic going on (e.g. some `&&`, `|`, for loops, etc.).

The current process handling checks for these two scenarios and sends signals accordingly. This is important as `sh` doesn't forward UNIX signals on its own, which sometimes results in dangling processes!

Other shells however vary wildly in their subprocess behaviors:
`fish`, for instance, likes to spawn other child processes such as python processes for handling man pages.
`fish` also spawns a `fish -c` instance under all circumstances.

We would have to account for all of these quirks in our process handling.
The best way to make sure that this works as expected would probably be to write dedicated process handling modules for each shell and only support a fix set of shells in pueue.

### Dedicated process handling logic per shell

The one downside I see with this, is that this significantly increases the maintenance overhead of the project.
It'll also lead to me no longer being able of debugging all of the logic used in the Unix code, which I managed to prevent since quite some time.

Testing this will also become very annoying.
One of pueue's core philosophies is to be independant of any external dependencies that aren't guaranteed to be there anyway.

The only shell that's guaranteed to exist is `sh`, which is present on any unix system.
In order to run tests, all developers would have to install all shells that're supported by pueue, which is very undesirable.

## How to make it configurable.

In the case of powershell, the contributor that added windows support made a template string for the command. The unix `sh` code works by passing the command as a OSstring directly, without any additional string handling.
As at least one of the shells need some kind of templating, we would have to make this possible under all circumstances.

I'm a bit wary of users working with string templates.
As soon as one works with quotes, one can very easily mess up string escaping.
For instance, `-c '{command}'` would make things spicy, as all commands with `'` inside of it would break in an interesting way.


Another thing I'm not sure about yet is how the configuration for this would look like:
This is what I came up with from the top of my head:
```toml
daemon:
    shell:
        bin: bash
        args:
            - '-rcfile=/etc/bash'
            - '-c'
        command:
            - "VARIABLE=test {}"
```

I assume that we have to come up with something better, as this is very error prone.
I would like this to be as foolproof as possible :D

------------------------------------------------------

Well, that's all of the stuff I can come up with for now that needs clarification or solutions.

As said at the top of this comment, I would really like to have this feature in Pueue.
I'm just wary due to my prior experiences (which to be fair were 5 years ago), and the problems I see with the process handling logic.

**Comment by WindSoilder** [comment]
Thanks for your reply, I think I understand your points and concern.

Could you please give me some advice if I just want to make `pueue` support for specific shell?  Actually what I want to do is making nushell's task management feature works like pueue. 

Here is my thought:
1. I think it's better to implement that based on `pueue-lib`.
2. For pueue-cli part, that's no problem, I need to implements some commands which is communicating to `pueued`.
3. For pueued part, I'm not sure if I really want to build a pueue-lib based, new nushell only `pueued`?  Or,  I just need to fork `pueue-daemon` code and adjust it to work with nushell ? (Because mostly of task handling part would be the same)

**Comment by Nukesor** [comment]
`pueue-lib` doesn't do much work on it's own.

Most of the beefy task management logic is contained in `pueued`.
`pueue-lib` only exposes common code used by both, the client and the daemon, such as the protocol, low-level process handling etc.

If you were to build your own fork of the pueue project, you might be better of to start with the `pueued` daemon and pick the pieces of code you actually need :)

**Comment by Nukesor** [comment]
Getting Pueue to **only** work with nushell would be relatively easy.

You would have to use the correct call, investigate how nushell spawns subprocess, handles Unix signals and adjust the process handling code accordingly.

If you want to add support for nushell on top of the current posix shell in the upstream pueue project, I'm not really sure how to do this yet. I sadly don't have a lot of time right now, but I'll try to come up with a good solution. This might take a bit of time though.

**Comment by WindSoilder** [comment]
Thank you very much, currently we've come up with different solution.  

> I'm not really sure how to do this yet. 

I'm not sure either, But we make it works (sending `nu -c xxx`, so pueued executes `sh -c "nu -c xxx"`)  I think the issue can be closed :-D
Thank you again for your help!  And I love `pueue` very much :-D

**Comment by Nukesor** [comment]
@WindSoilder 

Due to some internal refactorings, this has become much easier.

I'm currently working on an implementation of this feature in https://github.com/Nukesor/pueue/pull/458

If you're still interested, could you take a look at this :)?

**Comment by Nukesor** [comment]
Implemented in #458
